test/lib/test_runner.go

```diff
@@ -73,6 +74,47 @@ func (tr *ComponentsTestRunner) RunTests(
 	}
 }
 
+// RunTestsWithComponentOptions will use all components that support the given
+// feature, to run a test for the testFunc while passing the component specific
+// SetupClientOptions to testFunc. You should used this method instead of
+// RunTests if you have used AddComponentSetupClientOption to add some component
+// specific initialization code.
+func (tr *ComponentsTestRunner) RunTestsWithComponentOptions(
+	t *testing.T,
+	feature Feature,
+	testFunc func(st *testing.T, component metav1.TypeMeta, options ...SetupClientOption),
+) {
+	t.Parallel()
+	for _, component := range tr.ComponentsToTest {
+		// If a component is not present in the map, then assume it has all properties. This is so an
+		// unknown component (e.g. a Channel) can be specified via a dedicated flag (e.g. --channels) and have tests run.
+		// TODO Use a flag to specify the features of the flag based component, rather than assuming
+		// it supports all features.
+		features, present := tr.ComponentFeatureMap[component]
+		if !present || contains(features, feature) {
+			t.Run(fmt.Sprintf("%s-%s", component.Kind, component.APIVersion), func(st *testing.T) {
+				testFunc(st, component, tr.componentOptions[component]...)
```

Comment:
Wait but you don't run the setup function here right? I lost where that function runs :smile:

Reply:
I totally feel ya :disappointed: . The idea is, since `ComponentTestRunner` is just a selector of components, it includes components and their related ClientSetupOptions, and when running the tests, it passes the component and its ClientSetupOptions to the test function since the test function has the env context (the client it needs to inject those ClientSetupOptions with). I'm not a big fan of how convoluted the code ends up but that's the tension I've identified before between the helpers and the assumptions they make on those helpers consumers. 

If you have a cleaner way in mind to do this I'd gladly do it.

Reply:
Ok so the setup function of the source is a `ClientSetupOption`, which is ran when the client is brought up right?

Reply:
what really confuses me is that you're using the `ClientSetupOption` to setup the source. For me `[T]Option` fn in golang is a function that setups `T`, not that uses it... Maybe these setup functions should be run as "post client setup" functions?